### PR TITLE
Adding new ecn testcases for cisco-8000 platform.

### PR DESCRIPTION
**Pull request by rraghav-cisco**
New test case for cisco-8000 to demonstrate accuracy of ECN marking under continuous traffic when a single XOFF is injected.

Test is parameterized across several variables.

While waiting for IXIA to get number of ECN marked packets (Issue: [open-traffic-generator/snappi-ixnetwork#541](https://github.com/open-traffic-generator/snappi-ixnetwork/issues/541)), we are using the counters from DUT, using the extension added via ([#6040](https://github.com/sonic-net/sonic-mgmt/pull/6040)).

### CODE REVIEW

File: tests/ixia/ecn/files/helper.py [hunk 1]
```diff
@@ -40,10 +43,15 @@ def run_ecn_test(api,
                  kmax,
                  pmax,
                  pkt_size,
-                 pkt_cnt,
                  lossless_prio,
                  prio_dscp_map,
-                 iters):
+                 iters,
+                 pkt_cnt=NUMBER_OF_TEST_PACKETS,
```
Comment (baiwei0427) [inline tests/ixia/ecn/files/helper.py:49]: Can we rename this to data_pkt_cnt to differentiate with # of PFC packets?

File: tests/ixia/ecn/files/helper.py [hunk 2]
```diff
@@ -40,10 +43,15 @@ def run_ecn_test(api,
                  kmax,
                  pmax,
                  pkt_size,
```
Comment (baiwei0427) [inline tests/ixia/ecn/files/helper.py:45]: Rename this to data_pkt_size

File: tests/ixia/ecn/files/helper.py [hunk 3]
```diff
@@ -40,10 +43,15 @@ def run_ecn_test(api,
...
-                 pkt_cnt,
                  lossless_prio,
                  prio_dscp_map,
-                 iters):
+                 iters,
+                 pkt_cnt=NUMBER_OF_TEST_PACKETS,
+                 xoff_quanta=65535,
+                 traffic_rate=None,
```
Comment (baiwei0427) [inline tests/ixia/ecn/files/helper.py:51]: data_traffic_rate

File: tests/ixia/ecn/files/helper.py [hunk 4]
```diff
@@ -40,10 +43,15 @@ def run_ecn_test(api,
...
+                 number_of_transmit_ports=1,
+                 pause_delay=0,
+                 single_pause=False):
```
Comment (baiwei0427) [inline tests/ixia/ecn/files/helper.py:54]: We can rename this to pfc_cnt

File: tests/ixia/ecn/files/helper.py [hunk 5]
```diff
@@ -40,10 +43,15 @@ def run_ecn_test(api,
...
-                 pkt_cnt,
                  lossless_prio,
                  prio_dscp_map,
-                 iters):
+                 iters,
+                 pkt_cnt=NUMBER_OF_TEST_PACKETS,
+                 xoff_quanta=65535,
+                 traffic_rate=None,
+                 number_of_transmit_ports=1,
+                 pause_delay=0,
```
Comment (baiwei0427) [inline tests/ixia/ecn/files/helper.py:53]: pfc_storm_start_delay?

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

**Comment by baiwei0427** [comment]
My overall feeling is that current run_ecn_test is too complex. It has a mix of two mechanisms. Can we implement another run_ecn_test for Cisco only?

File: tests/ixia/ecn/files/helper.py (hunk 1 above)
Comment (rraghav-cisco) [reply tests/ixia/ecn/files/helper.py:49]: Done.

**Comment by rraghav-cisco** [comment]
> My overall feeling is that current run_ecn_test is too complex. It has a mix of two mechanisms. Can we implement another run_ecn_test for Cisco only?

@baiwei0427 : Thanks for your comments. I started with seperate functions one each for 1-1 and 3-1 cases, and then combined them all into the original function. There was a lot of similarity between them and the original function, so it was decided to merge them all.

File: tests/ixia/ecn/files/helper.py [hunk 6]
```diff
@@ -98,42 +114,60 @@ def run_ecn_test(api,
     pytest_assert(port_id is not None,
                   'Failed to get ID for port {}'.format(dut_port))
 
-    """ Generate packet capture config """
-    capture_config = __config_capture_ip_pkt(testbed_config=testbed_config, port_id=port_id)
+    capture_config = None
```
Comment (baiwei0427) [inline tests/ixia/ecn/files/helper.py:117]: I think we should add a parameter 'enable_capture' for the function

File: tests/ixia/ecn/files/helper.py [hunk 7]
```diff
@@ -98,42 +114,60 @@ def run_ecn_test(api,
...
+    else:
+        for i in range(number_of_transmit_ports):
+            name_options['data_flow_names'].append(DATA_FLOW_NAME + " " + str(i+1))
     """ Generate traffic config """
     flows = __gen_traffic(testbed_config=testbed_config,
                           port_config_list=port_config_list,
                           port_id=port_id,
                           pause_flow_name=PAUSE_FLOW_NAME,
-                          data_flow_name=DATA_FLOW_NAME,
                           prio=lossless_prio,
```
Comment (baiwei0427) [inline tests/ixia/ecn/files/helper.py:134]: Is parameter data_flow_names missing?

File: tests/ixia/ecn/files/helper.py [hunk 8]
```diff
@@ -249,24 +303,33 @@ def __gen_traffic(testbed_config,
...
+                                           delay=pfc_storm_start_delay,
+                                           delay_unit='nanoseconds')))
+    else:
+        pause_flow = Flow(
+            name=pause_flow_name,
+            tx_rx=TxRx(pause_endpoint),
+            packet=[pause_pkt],
+            size=Size(64),
+            rate=Rate('pps', value=pps),
+            duration=Duration(FixedSeconds(seconds=exp_dur_sec,
```
Comment (baiwei0427) [inline tests/ixia/ecn/files/helper.py:326]: Align arguments

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/ixia/ecn/files/helper.py (hunk 7 above)
Comment (rraghav-cisco) [reply tests/ixia/ecn/files/helper.py:134]: It is part of "name_options".

**Comment by rraghav-cisco** [comment]
@baiwei0427 : Pls let me know if you have comments on the latest code.

**Comment by baiwei0427** [comment]
> @baiwei0427 : Pls let me know if you have comments on the latest code.

@rraghav-cisco I will add comments asap.

File: tests/ixia/ecn/test_ecn_markings_cisco_8000.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
+import pytest
+import time
+import datetime
+import  os.path
```
Comment (baiwei0427) [inline tests/ixia/ecn/test_ecn_markings_cisco_8000.py:4]: Fix the space between import and os.path

File: tests/ixia/ecn/test_ecn_markings_cisco_8000.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        	 lat=latency_marking,
+        	 data_pkt_size=data_pkt_size,
+        	 tx_ports=number_of_transmit_ports,
+        	 kmax=kmax,
+        	 kmin=kmin,
+        	 pmax=pmax,
+        	 xoff_quanta=xoff_quanta,
+        	 total=total,
+        	 marked=marked_pkts))
+    print 'test done'
```
Comment (baiwei0427) [inline tests/ixia/ecn/test_ecn_markings_cisco_8000.py:161]: It seems that there is no logic to judge if the test passes.

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/ixia/ecn/test_ecn_markings_cisco_8000.py (hunk 10 above)
Comment (rraghav-cisco) [reply tests/ixia/ecn/test_ecn_markings_cisco_8000.py:161]: The test is passed if run_ecn_test is successfully done. That function itself has the required asserts to check.

Comment (baiwei0427) [reply tests/ixia/ecn/test_ecn_markings_cisco_8000.py:161]: > The test is passed if run_ecn_test is successfully done. That function itself has the required asserts to check.

Where are the required asserts in run_ecn_test?

Comment (rraghav-cisco) [reply tests/ixia/ecn/test_ecn_markings_cisco_8000.py:161]: @baiwei0427 , I was referring to the following checks. If any of them fails, the test will fail.

grep -n assert files/helper.py 
4:from tests.common.helpers.assertions import pytest_assert
89:    pytest_assert(testbed_config is not None, 'Failed to get L2/3 testbed config')
99:    pytest_assert(config_result is True, 'Failed to configure WRED/ECN at the DUT')
108:    pytest_assert(config_result is True, 'Failed to configure PFC threshold to 8')
116:    pytest_assert(port_id is not None,
221:    pytest_assert(len(tx_port_id_list) > 0, "Cannot find any TX ports")
224:    pytest_assert(tx_port_id_list != [], "Cannot find a suitable TX port")
396:    pytest_assert(attempts < max_attempts,

**Comment by rraghav-cisco** [comment]
This needs to be merged after 6329 is merged.

**Comment by rraghav-cisco** [comment]
@kevinskwang Any update on azp failures ?

**Comment by kevinskwang** [comment]
Could you fix the conflict by first?

**Comment by rraghav-cisco** [comment]
> Could you fix the conflict by first?

@kevinskwang : Conflict is resolved. Pls check.

**Review by kevinskwang — APPROVED** [review]

**Comment by kevinskwang** [comment]
Reverted as some import errors, please fix it and create a new PR for this. @rraghav-cisco

**Comment by rraghav-cisco** [comment]
> Reverted as some import errors, please fix it and create a new PR for this. @rraghav-cisco

@kevinskwang , sure. I will take it up after current priorities. Thanks.

**Comment by developfast** [comment]
@rraghav-cisco Currently, we are not running any test cases in the ixia directory. All of our RDMA related test cases have been migrated to the snappi directory here. The new ecn test would go in this directory - https://github.com/sonic-net/sonic-mgmt/tree/master/tests/snappi/ecn. Please make the new code changes to this directory when you bring up the new PR. Thanks!

**Pull request merged**